File: android/src/main/kotlin/video/api/flutter/livestream/CameraSettingsHostApiImpl.kt
package video.api.flutter.livestream

import io.github.thibaultbee.streampack.core.elements.sources.video.camera.CameraSettings
import io.github.thibaultbee.streampack.core.elements.sources.video.camera.ICameraSource
import kotlinx.coroutines.runBlocking
import video.api.flutter.livestream.generated.CameraSettingsHostApi
import video.api.flutter.livestream.manager.InstanceManager

class CameraSettingsHostApiImpl(
    private val instanceManager: InstanceManager
) :
    CameraSettingsHostApi {
    private val settings: CameraSettings
        get() {
            val streamer = runBlocking { instanceManager.getInstance() }
            return (streamer.videoInput.sourceFlow.value as? ICameraSource)?.settings
                ?: throw IllegalStateException("Camera source is not available")
        }

    override fun setZoomRatio(zoomRatio: Double) {
        runBlocking { settings.zoom.setZoomRatio(zoomRatio.toFloat()) }
    }

    override fun getZoomRatio(): Double {
        return runBlocking { settings.zoom.getZoomRatio().toDouble() }
    }
}

File: ios/apivideo_live_stream/Sources/apivideo_live_stream/CameraSettingsHostApiImpl.swift
import AVFoundation

class CameraSettingsHostApiImpl: CameraSettingsHostApi {
    let instanceManager: InstanceManager

    init(instanceManager: InstanceManager) {
        self.instanceManager = instanceManager
    }

    func setZoomRatio(zoomRatio: Double) throws {
        MainActor.assumeIsolated {
            guard let liveStream = instanceManager.liveStream else {
                return
            }
            liveStream.zoomRatio = CGFloat(zoomRatio)
        }
    }

    func getZoomRatio() throws -> Double {
        MainActor.assumeIsolated {
            guard let liveStream = instanceManager.liveStream else {
                return 1.0
            }
            return Double(liveStream.zoomRatio)
        }
    }
}
